### PR TITLE
Add Sub(a,b) [subtract] and Intersect(a,b) operators to sedlex%matches

### PR DESCRIPTION
**Pull request by mcclure**
...Both ops work only on character sets. Also fix up a couple comments and confusing error language on Compl.

My use case is something like Sub(white_space, "\n"). I am not sure why anyone would want Intersect. It seems like it could be useful very occasionally.

Here's my test program for the additions:

(* Simple sedlex sample *)

      let () = 
            let rec lex buf = 
                  match%sedlex buf with
                        | white_space -> print_endline "\tWhitespace"; lex buf
                        | Sub (Chars "ab","b") -> print_endline "a"; lex buf
                        | (Chars "ab"|"c") -> print_endline "abc"; lex buf
                        | Intersect ("d", Chars "abd") -> print_endline "d"; lex buf
                        | eof -> print_endline "\tEnd"
                        | any -> print_endline "Other"; lex buf
                        | _ -> failwith "Internal failure: Reached impossible place"
            in lex @@ Sedlexing.Utf8.from_string "a b c d e"

First time doing something like this in the ocaml community, let me know if I did anything suboptimal :O

### CODE REVIEW

**Comment by whitequark** [comment]
LGTM. /cc @Drup

File: src/syntax/ppx_sedlex.ml [hunk 1]
```diff
@@ -282,8 +289,12 @@ let regexp_of_pattern env =
         | Some r -> r
         | None ->
           err p.ppat_loc
-            "the Compl operator can only applied to a single-character regexp"
+            "the Compl operator can only applied to a single-character length regexp"
         end
+    | Ppat_construct ({ txt = Lident "Sub" }, Some {ppat_desc=Ppat_tuple (p0 :: p1 :: [])}) ->
+        char_pair_op Sedlex.subtract "Sub" p p0 p1
+    | Ppat_construct ({ txt = Lident "Intersect" }, Some {ppat_desc=Ppat_tuple (p0 :: p1 :: [])}) ->
+        char_pair_op Sedlex.intersection "Intersect" p p0 p1
```
Comment (Drup) [inline src/syntax/ppx_sedlex.ml]: I wouldn't mind a sensible error message for wrong arity here.

File: src/syntax/ppx_sedlex.ml [hunk 2]
```diff
@@ -282,8 +289,12 @@ let regexp_of_pattern env =
         | Some r -> r
         | None ->
           err p.ppat_loc
-            "the Compl operator can only applied to a single-character regexp"
+            "the Compl operator can only applied to a single-character length regexp"
```
Comment (Drup) [inline src/syntax/ppx_sedlex.ml]: The same expression is used multiple times in the doc, can you change it there as well ?

File: src/syntax/sedlex.mli [hunk 3]
```diff
@@ -12,6 +12,8 @@ val plus: regexp -> regexp
 val eps: regexp
 
 val compl: regexp -> regexp option
+val subtract: regexp -> regexp -> regexp option
+val intersection: regexp -> regexp -> regexp option
    (* If the argument is a single [chars] regexp, returns a regexp
       which matches the complement set.  Otherwise returns [None]. *)
```
Comment (Drup) [inline src/syntax/sedlex.mli]: The comment is not in the right place anymore (and you can add other comments for the two new functions).

**Comment by Drup** [comment]
The code is fine and works afaics. Thanks! 

Can you fix the various comments and add some documentation in the readme ?

**Comment by mcclure** [comment]
Cool, will do.

**Comment by mcclure** [comment]
I believe the new commit covers all the issues raised above, I also added arity checks for Compl and Chars.

**Comment by Drup** [comment]
@mcclure Do I have your authorization to add your sample in the example directory ?

**Comment by mcclure** [comment]
Drup: Yes.

**Pull request merged**